store/tikv/region_cache_test.go

```diff
@@ -259,22 +311,46 @@ func (s *testRegionCacheSuite) TestReconnect(c *C) {
 	s.checkCache(c, 1)
 }
 
-func (s *testRegionCacheSuite) TestRequestFail(c *C) {
+func (s *testRegionCacheSuite) TestSendFailBlackout(c *C) {
+	ts := time.Now().Unix()
 	region := s.getRegion(c, []byte("a"))
 
-	ctx, _ := s.cache.GetRPCContext(s.bo, region.VerID())
-	s.cache.DropStoreOnSendRequestFail(ctx, errors.New("test error"))
-	c.Assert(s.cache.mu.regions, HasLen, 0)
-	region = s.getRegion(c, []byte("a"))
-	c.Assert(s.cache.mu.regions, HasLen, 1)
-	ctx, _ = s.cache.GetRPCContext(s.bo, region.VerID())
-	s.cache.DropStoreOnSendRequestFail(ctx, errors.New("test error"))
-	c.Assert(len(s.cache.mu.regions), Equals, 0)
+	// init with 1 region 2 stores
+	c.Assert(workableRegions(s.cache, s.cache.mu.regions, ts), Equals, 1)
+	c.Assert(reachableStore(s.cache.storeMu.stores, ts), Equals, 2)
+
+	// for each stores has 20 chance to retry, and still have chance to access store for 21
+	for i := 0; i < 38; i++ {
+		ctx, _ := s.cache.GetRPCContext(s.bo, region.VerID())
+		if ctx == nil {
+			fmt.Println()
+		}
+		s.cache.OnSendRequestFail(ctx, errors.New("test error"))
+
+	}
+	ts = time.Now().Unix()
+	c.Assert(workableRegions(s.cache, s.cache.mu.regions, ts), Equals, 1)
+	c.Assert(reachableStore(s.cache.storeMu.stores, ts), Equals, 2)
+
+	// 21 fail attempts will start blackout store in 1 second
+	for i := 0; i < 2; i++ {
+		// first fail request make 1st store' failAttempt + 1
+		ctx, _ := s.cache.GetRPCContext(s.bo, region.VerID())
+		s.cache.OnSendRequestFail(ctx, errors.New("test error"))
+	}
+	c.Assert(workableRegions(s.cache, s.cache.mu.regions, ts), Equals, 0)
+	c.Assert(reachableStore(s.cache.storeMu.stores, ts), Equals, 0)
+
+	// after 1 second blackout, 2 store can be access again.
+	time.Sleep(1 * time.Second)
```

Comment:
It would make CI slower, is there any way to avoid that?

Reply:
this is timeaware test, we can mock time to acc it, but maybe high price to do that..